### DISCUSSION TITLE
关于 state 的中文翻译使用讨论

### DISCUSSION THREAD

**Discussion by awxiaoxian2020** [reactjs/zh-hans.react.dev#1126] [Ideas]
首先我们需要说明以下几个事实（如果有误，欢迎指出）

- “状态”这个词在计算机、社科以及其他自然科学里面都有非常多的确切的含义。不像“组件”这个词在别的领域并不常见，因为 component 这个词本来的译法是“组成部分”，“组件”在一定程度上和范围内已经是前端框架的专有名词了。
- 日常开发交流中，使用英文单词较多
- “state”是 React 为了管理状态有时会使用的一个专有名词
- “state”翻译成“状态”，在翻译结果的表意上和原文是一致的，只是作翻译会带来表意泛化，不够精确的问题

以下是我的一些粗略的想法：

---

- 存在代码上下文，且有关内容和代码强关联时，“state”一律不译。这可以使读者迅速想到这是 React 的专有名词，和代码关联起来。
- 有其他不译的单词在同一句话里面互为上下文时一律不译（如 `props 和 state`），因为就算翻译，也需要加上双引号表特殊含义吧。

---

- 在其作为定语时，考虑翻译。如 `状态管理` `状态提升`，因为这里的表意十分明确，在前端的文档里面，不是随便什么“状态”都能够或者说需要管理。
- 在讨论原理和设计哲学时，应在不与前两条冲突的情况下予以翻译。因为这种情况已经脱离具体实现，而将其作为一种良好的设计模式或解决方案，此时上下文已经限制了表意的范围，语义明确，翻译也可使这种思想更好地推广出去。

**Comment by awxiaoxian2020** [comment]
对外发布规范时，表述需要简化。

**Comment by fwqaaq** [comment]
>这可以使读者迅速想到这是 React 的专有名词，和代码关联起来。

这并不是 react 所专有的，只是在使用 class 才会有的独有属性，在 vue 中同样也有，它们都代表组件 UI 状态的变化。

**Reply by awxiaoxian2020** [reply]
请考虑我提出的应当翻译的情形~

**Reply by fwqaaq** [reply]
> 存在代码上下文，且有关内容和代码强关联时，“state”一律不译。这可以使读者迅速想到这是 React 的专有名词，和代码关联起来。
>
> “state”是 React 为了管理状态有时会使用的一个专有名词

这并不是 react 独有的，vue、solid....但凡响应式的框架都会有状态这个概念，而且响应式（stateful）这个词根就是 state

>其他不译的单词在同一句话里面互为上下文时一律不译（如 props 和 state），因为就算翻译，也需要加上双引号表特殊含义吧。

你可以参考 vue3 以及 solid 等框架是如何理解 state 这个概念是否翻译

> “状态”这个词在计算机、社科以及其他自然科学里面都有非常多的确切的含义。
>
>在其作为定语时，考虑翻译。如 状态管理 状态提升，因为这里的表意十分明确，在前端的文档里面，不是随便什么“状态”都能够或者说需要管理。
>
> “state”翻译成“状态”，在翻译结果的表意上和原文是一致的，只是作翻译会带来表意泛化，不够精确的问题

在前端中，"状态"通常指的是应用程序或用户界面中的数据或信息的当前值或状态。并不是说什么“状态”都能够或者说需要管理，而是在前端中这个含义很明确。

1. 组件状态：在前端框架（如 React、Vue 等）中，组件是构建用户界面的基本单元，每个组件都可以拥有自己的状态。
2. 应用程序状态：整个应用程序的状态，通常由多个组件共享或管理。
3. 用户界面状态：用户界面的显示状态，例如当前选中的标签、展开/折叠的面板、显示/隐藏的元素等。

之前的 `props 和 state` 举例，state 是引用的这部分 <https://react.dev/learn/tutorial-tic-tac-toe#making-an-interactive-component>，其就是在讲组件 UI 状态。

关于“表意泛化，不够精确”，很显然如果仅使用 `state` 这个单词，如果是初学者也是同样的道理

> 日常开发交流中，使用英文单词较多

不确切含义的单词使用英文没问题，如果表意明确、有专业术语的可以不使用

**Reply by awxiaoxian2020** [reply]
关于你所说的，状态这个词在前端中很常见，表意是明确的这一点，我拿 ChatGPT 生成了“状态”一词可能对应的英文单词：
```markdown
    State
    **Status**
    Condition
    **Situation**
    Position
    Mode
    Phase
    **Stage**
    Form
    Level
```
至少在我加粗的这几个词语里面，可能出现在前端的文档里面。我们不必争辩它们是否真的应该被翻译为“状态”，但是出现混用的风险还是有的，我们也不能穷举可能混用的情况，诚然我们可以通过 review 来解决这个问题，但是需要耗费的精力和讨论成本是不可比的。

>组件状态：在前端框架（如 React、Vue 等）中，组件是构建用户界面的基本单元，每个组件都可以拥有自己的状态。
应用程序状态：整个应用程序的状态，通常由多个组件共享或管理。
用户界面状态：用户界面的显示状态，例如当前选中的标签、展开/折叠的面板、显示/隐藏的元素等。

就你提出的这些段落和上下文，我个人也是赞同不译的，这对应的是我刚刚说的原理和设计哲学，它放在任何一个地方都是可以的。正如你所说，一些前端的框架都有状态的概念，这和 React 是无关的。

> 日常开发交流中，使用英文单词较多

我再举个例子，你的发言中使用了 class，那为什么不翻译成“类”相关术语呢？另外摘自 Vue 的翻译指南:

> class JS 中翻译成“类”，CSS 中不翻译

UI 翻译成用户界面表意也没有问题，为什么日常生活中既有说 UI 也有说用户界面的情景呢？

再比如我自己的这段话：

> 诚然我们可以通过 review 来解决这个问题，但是需要耗费的精力和讨论成本是不可比的。

为什么我会不自觉地使用 review 而不是审查呢？我刚刚想到，正是因为我们天天看见 GitHub 上面的 review 按钮，自然会用它作为代指。审查放在这里可能就有歧义了。当然我如果说“代码审查”可能就避免了这个问题（也不一定……），但是你不能因为这就否认**使用英文单词的合理性**。

另外如果非要拿其他的框架翻译作为参考的话，[pinia](https://pinia.vuejs.org/zh/core-concepts/state.html) 在阐述 state 和 store 时并没有翻译，如果都需要翻译，那么下面这句话就比较尴尬了：

> In Pinia the state is defined as a function that returns the initial state.

还有：

> 虽然这在简单的情况下是可行的，但从长远来看，可以被任何组件任意改变的全局状态是不太容易维护的。为了确保改变状态的逻辑像状态本身一样集中，建议在 store 上定义方法，方法的名称应该要能表达出行动的意图：

store 在这里的表意也是明确的，但翻译出来就比较奇怪。

我的表达能力有限，但我觉得是否翻译的根本宗旨还是要符合现有生态的习惯。

**Reply by fwqaaq** [reply]
pinia 那点是因为它的 state 是固有属性，就像你之前说的对于引用的代码是不翻译的。然而很多明明能表意的语句，现在读起来就像，我明明可以 speak 中文，但是我现在就要说 English

**Reply by awxiaoxian2020** [reply]
看来我们其实是有共识的。但是这个是否翻译的边界实在是难以确定（至少我不能完美表达出来），要不换一个讨论方向，什么情况下应该不翻译或者说现有的不翻译的情况是怎样的。

**Reply by fwqaaq** [reply]
是，或许应该罗列一些示例，🤔

**Reply by awxiaoxian2020** [reply]
其实这是十分灵活的，比如我赞同 `组件和状态` 但是我也支持“组件、props 和 state”而不是“组件、状态和 props”……

组件一词并没有在 API 中出现很多次，更多地是作为一个容器（表述不太准确），英文很少碰到，但是 state 由于有像 useState() 以及其他的代码中经常会出现，我问过一些开发人员，他们都表示 state 这个单词在日常交流中使用较多，所以你如果没有方案我提出的这些使用现象和问题，或者是给出清晰的是否翻译的标准，那我们还是采用多不翻译的策略。

也欢迎你提出一些可供操作的，不会有歧义的是否翻译的方案。

**Reply by fwqaaq** [reply]
好吧，先这样

**Comment by Huxpro** [comment]
@awxiaoxian2020 我基本同意你的看法，有没有可能总结一个 guideline 来指导以后的 contributors 什么时候翻译什么时候不翻呢？

**Comment by Yucohny** [comment]
See https://github.com/reactjs/zh-hans.react.dev/issues/1354.
